examples/step-a/step-a.cc

```diff
@@ -0,0 +1,1654 @@
+        int(parameters.geometry.dimension_x / parameters.geometry.dimension_y);
+      divisions[1] = 1;
+      if (dim == 3)
+        {
+          divisions[2] = 1;
+        }
+      GridGenerator::subdivided_hyper_rectangle(triangulation,
+                                                divisions,
+                                                p1,
+                                                p2);
+    }
+
+    triangulation.refine_global(parameters.geometry.grid_level);
+
+    // Run the frequency sweep
+    frequency_sweep();
+  }
+
+  template <int dim, typename ParametersT>
+  void ElasticWave<dim, ParametersT>::frequency_sweep()
+  {
+    // This data will be used by store_frequency_step_data()
+    store_position_data();
+    for (unsigned int frequency_idx = 0;
+         frequency_idx < parameters.nb_frequency_points;
+         ++frequency_idx)
+      {
+        std::cout << parameters.simulation_name + " frequency idx: "
+                  << frequency_idx << '/' << parameters.nb_frequency_points - 1
+                  << std::endl;
+
+
+
+        setup_system();
+        if (frequency_idx == 0)
+          {
+            std::cout << "   Number of active cells :       "
+                      << triangulation.n_active_cells() << std::endl;
+            std::cout << "   Number of degrees of freedom : "
+                      << dof_handler.n_dofs() << std::endl;
+          }
+
+        if (frequency_idx == 0)
+          {
+            // Write the simulation parameters only once
+            parameters.data.write_int_attr("active_cells",
+                                           triangulation.n_active_cells());
+            parameters.data.write_int_attr("degrees_of_freedom",
+                                           dof_handler.n_dofs());
+            parameters.data.write_int_attr("dimensions", dim);
+          }
+
+        double current_loop_frequency =
+          (parameters.start_frequency +
+           frequency_idx *
+             (parameters.stop_frequency - parameters.start_frequency) /
+             (parameters.nb_frequency_points - 1));
+        double current_loop_omega = 2 * numbers::PI * current_loop_frequency;
+        assemble_system(current_loop_omega);
+        solve();
+
+        frequency[frequency_idx] = current_loop_frequency;
+        store_frequency_step_data(frequency_idx);
+      }
+    output_results();
+  }
+
+  template <int dim>
+  SymmetricTensor<4, dim> get_stiffness_tensor(const double lambda,
+                                               const double mu)
+  {
+    SymmetricTensor<4, dim> stiffness_tensor;
+    for (unsigned int i = 0; i < dim; ++i)
+      for (unsigned int j = 0; j < dim; ++j)
+        for (unsigned int k = 0; k < dim; ++k)
+          for (unsigned int l = 0; l < dim; ++l)
+            stiffness_tensor[i][j][k][l] =
+              (((i == k) && (j == l) ? mu : 0.0) +
+               ((i == l) && (j == k) ? mu : 0.0) +
+               ((i == j) && (k == l) ? lambda : 0.0));
```

Comment:
This is not wrong, but I suspect that you can write this in simpler form using the predefined rank-4 tensors either in the physics module or the ones that are marked as `@related` to the `SymmetricTensor<4,dim>` type.

But it may be instructive to actually see this here. When you write the documentation, you can always refer to using the other functions.

Reply:
You are right! I think that I can use the double_contract operation from the Tensor class.
The tensor is non-symmetrical because of the PMLs.